### PR TITLE
Add summary of our coding conventions

### PR DESCRIPTION
**Pull request by PaulWessel**
This document starts to outline the naming conventions used for functions, structures, macros, and constants (including enums) in GMT.

### CODE REVIEW

File: GMT_CODE_CONVENTIONS.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
+# GMT Code Conventions
+
+## Contents
+
+* [The documented external GMT API](#the-api)
+* [The undocumented internal GMT library](#the-Developer-library)
```
Comment (seisman) [inline GMT_CODE_CONVENTIONS.md]:
```suggestion
* [The undocumented internal GMT library](#the-developer-library)
```

File: GMT_CODE_CONVENTIONS.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+library.  To keep these as organized as possible, we need follow our naming
+guidelines.  As of this writing, we are violating these guidelines in places
+and will seek to reach a better level of compliance as we go forward.
+
+## The API
+
+The official API provides declaration for all items accessible by
+the users of the API (e.g., the external interfaces to GMT from
+MATLAB, Python, and Julia). The are all accessed by including
+[gmt.h](https://github.com/GenericMappingTools/gmt/blob/master/src/gmt.h)
```
Comment (seisman) [inline GMT_CODE_CONVENTIONS.md]: No need to give the full URL.
```suggestion
[gmt.h](src/gmt.h)
```

File: GMT_CODE_CONVENTIONS.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+There is only one data type defined in the API and that is *gmt_grdfloat*.  This
+is typeset to *float* but developers with special needs can change this to *double*.
+This means all GMT grids would use double precision [default is float].
+
+## The Developer Library
+
+Developers writing in C or C++ may choose to include [gmt_dev.h](https://github.com/GenericMappingTools/gmt/blob/master/src/gmt_dev.h)
+in their codes. This opens up access to a myriad of additional, lower-level functions,
+macros, structures, and enums and constants.  These are not documented and
+probably will never be documented.  however, we do have naming conventions
```
Comment (seisman) [inline GMT_CODE_CONVENTIONS.md]:
```suggestion
probably will never be documented.  However, we do have naming conventions
```

File: GMT_CODE_CONVENTIONS.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+There is only one data type defined in the API and that is *gmt_grdfloat*.  This
+is typeset to *float* but developers with special needs can change this to *double*.
+This means all GMT grids would use double precision [default is float].
+
+## The Developer Library
+
+Developers writing in C or C++ may choose to include [gmt_dev.h](https://github.com/GenericMappingTools/gmt/blob/master/src/gmt_dev.h)
+in their codes. This opens up access to a myriad of additional, lower-level functions,
+macros, structures, and enums and constants.  These are not documented and
```
Comment (seisman) [inline GMT_CODE_CONVENTIONS.md]: Warn that these functions are unstable and may change at any time.

File: GMT_CODE_CONVENTIONS.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+a low-level function in gmt_support.c that deals with parsing any fill
+argument containing a pattern is called **gmtsupport_parse_pattern**.  This
+function cannot be called outside gmt_support.c.  This naming convention makes
+it easy to find functions when debugging is needed.
+
+### GMT C macros
+
+GMT defines numerous C macros for things were a functions would be overkill, given
+the extra overhead in calling a function.  Again, access to all GMT macros is given by
+including gmt_dev.h](https://github.com/GenericMappingTools/gmt/blob/master/src/gmt_dev.h).
```
Comment (seisman) [inline GMT_CODE_CONVENTIONS.md]:
```suggestion
including [gmt_dev.h](https://github.com/GenericMappingTools/gmt/blob/master/src/gmt_dev.h).
```

**Comment by seisman** [comment]
Maybe change the file name to "CODE_CONVECTIONS.md", as we also know that this file is for GMT.

**Comment by PaulWessel** [comment]
I should have said I think this is too long to be in the general contributor file - it may get longer as well.

**Comment by seisman** [comment]
I agree, but could you rename the file to "CODE_CONVENTIONS.md"?

**Comment by PaulWessel** [comment]
Sure can.

**Pull request merged**